pkg/kstatus/watcher/watcher_dynamic.go

```diff
@@ -0,0 +1,825 @@
+			handleError(eventCh, fmt.Errorf("failed to compute object status: %s: %w", id, err))
+			return
+		}
+
+		if object.IsCRD(obj) {
+			w.onCRDUpdate(ctx, obj)
+		}
+
+		if isUnschedulable(rs) {
+			// schedule delayed failure & save the cancelFunc to cancel if there's an update
+			w.delayedEventCanceller.Add(id, w.scheduleFailure(ctx, eventCh, id))
+		}
+
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: rs,
+		}
+	}
+
+	handler.DeleteFunc = func(iobj interface{}) {
+		obj, ok := iobj.(*unstructured.Unstructured)
+		if !ok {
+			panic(fmt.Sprintf("DeleteFunc received unexpected object type %T", iobj))
+		}
+		id := object.UnstructuredToObjMetadata(obj)
+		if w.ObjectFilter.Filter(obj) {
+			klog.V(7).Infof("DeleteFunc: Watch Event Skipped: %s", id)
+			return
+		}
+		klog.V(5).Infof("DeleteFunc: Computing status for object: %s", id)
+
+		// Cancel any delayed events for this object
+		w.delayedEventCanceller.Cancel(id)
+
+		if object.IsCRD(obj) {
+			w.onCRDDelete(obj)
+		}
+
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: deletedStatus(id),
+		}
+	}
+
+	return handler
+}
+
+// onCRDAdd handles creating a new informer to watch the new resource type.
+func (w *ObjectStatusInformer) onCRDAdd(ctx context.Context, obj *unstructured.Unstructured) {
+	groupKind, found := object.GetCRDGroupKind(obj)
+	if !found {
+		id := object.UnstructuredToObjMetadata(obj)
+		klog.Warningf("Invalid CRD added: missing group and/or kind: %v", id)
+		// Don't return an error, because this should not inturrupt the task queue.
+		// TODO: Allow non-fatal errors to be reported using a specific error type.
+		return
+	}
+	klog.V(3).Infof("CRD added for %s", groupKind)
+
+	klog.V(3).Info("Resetting RESTMapper")
+	// Reset mapper to invalidate cache.
+	meta.MaybeResetRESTMapper(w.Mapper)
+
+	if w.pendingWatchers.Pop(groupKind) {
+		// Start a new informer for the GroupKind registered by this CRD.
+		err := w.startForGroupKind(
+			ctx,
+			groupKind,
+		)
+		if err != nil {
+			// Re-add to retry on next CRD update
+			w.pendingWatchers.Put(groupKind)
+			klog.Warningf("Failed to start watch for %v (will retry on next update): %v", groupKind, err)
+		}
+	}
+}
+
+// onCRDUpdate handles creating a new informer to watch the updated resource type.
+func (w *ObjectStatusInformer) onCRDUpdate(ctx context.Context, obj *unstructured.Unstructured) {
+	groupKind, found := object.GetCRDGroupKind(obj)
+	if !found {
+		id := object.UnstructuredToObjMetadata(obj)
+		klog.Warningf("Invalid CRD updated: missing group and/or kind: %v", id)
+		// Don't return an error, because this should not inturrupt the task queue.
+		// TODO: Allow non-fatal errors to be reported using a specific error type.
+		return
+	}
+	klog.V(3).Infof("CRD updated for %s", groupKind)
+
+	klog.V(3).Info("Resetting RESTMapper")
+	// Reset mapper to invalidate cache.
+	meta.MaybeResetRESTMapper(w.Mapper)
+
+	if w.pendingWatchers.Pop(groupKind) {
```

Comment:
What happens if there is already a watcher for this groupKind? Do we need to restart it?

Reply:
If there's already an informer then the GK doesn't get added to `pendingWatchers`. It only gets added to `pendingWatchers` if the CRD doesn't exist yet or if the previous informer has been deleted. No informer instance is restarted, because they're not designed to support that. A new informer instance is created, if needed (ex: CRD deleted and then re-created).

Reply:
If an informer exist for a GK and the CRD for this type is updated, does the informer get affected?

Reply:
Not currently. Presumably the existing informer will continue receiving updates, using whichever GVK it was already receiving. If the CRD was updated without a version bump, the changes should be reflected automatically. If the CRD version was incremented, the old default version would continue to be received.

I haven't tested for this, but this is what I assume would happen based on what i know of how the informers work.

Do you think this behavior will be problematic?

Reply:
I could add code to replace the informer if a new GVK is added by CRD update. I'm not sure how to prevent dropping events or producing duplicate events in that case tho.

Reply:
> When the CRD version is changed from "v1alpha" to "v1beta1" and "v1alpha1" is not supported. If the informer is for "v1alpha1" version, I'm not sure if it receives events when the version is changed to "v1beta1".

IME, when a new version is added, the old version usually sticks around for a while, rather than being immediately deleted. This allows CRs to be migrated from old to new. Otherwise, you'd have to delete all the old CRs before replacing "v1alpha1" with "v1beta1". When this happens, the old version should continue to work, it just won't necessarily be the default version. AFAIK, all installed versions need to be convertible to and from the default versions. So even if a controller upgrades to use a new "v1beta1" version, clients using the old "v1alpha1" version should still works. And that's what happening here. The Informer is a client using a specific version and that should keep working until that version is deleted from the CRD.

However, I agree that we should probably handle schema updates. And I agree that can be added in a future PR.

I suspect that the ideal behavior for a general use watcher impl would be to have the user specify which version they want to watch, but this won't be possible with the current API using ObjMetadata. The best we can do when only specifying group & kind as input is to always return the object using the cluster default version. I think it should be possible to do this with the existing informers, but it's going to require some client-side event de-duplication in order to avoid dropping events when migrating between informers for old and new versions. I suspect this probably requires a new informer wrapper that takes GK and manages the GVK itself.